File: src/main/java/xyz/oribuin/skyblock/managers/ConfigManager.java
package xyz.oribuin.skyblock.managers;

import org.bukkit.configuration.file.FileConfiguration;
import xyz.oribuin.skyblock.Skyblock;

import java.util.List;

public class ConfigManager extends Manager {

    public ConfigManager(Skyblock plugin) {
        super(plugin);
        this.reload();
    }

    public enum Setting {
        WORLD_NAME("world-name"),

        ISLAND_SETTINGS_SIZE("island-settings.size");

        private final String key;
        private Object value = null;

        Setting(String key) {
            this.key = key;
        }

        /**
         * Gets the setting as a boolean
         *
         * @return The setting as a boolean
         */
        public boolean getBoolean() {
            return (boolean) this.value;
        }

        /**
         * @return the setting as an int
         */
        public int getInt() {
            return (int) this.getNumber();
        }

        /**
         * @return the setting as a long
         */
        public long getLong() {
            return (long) this.getNumber();
        }

        /**
         * @return the setting as a double
         */
        public double getDouble() {
            return this.getNumber();
        }

        /**
         * @return the setting as a float
         */
        public float getFloat() {
            return (float) this.getNumber();
        }

        /**
         * @return the setting as a String
         */
        public String getString() {
            return (String) this.value;
        }

        private double getNumber() {
            if (this.value instanceof Integer) {
                return (int) this.value;
            } else if (this.value instanceof Short) {
                return (short) this.value;
            } else if (this.value instanceof Byte) {
                return (byte) this.value;
            } else if (this.value instanceof Float) {
                return (float) this.value;
            }

            return (double) this.value;
        }

        /**
         * @return the setting as a string list
         */
        @SuppressWarnings("unchecked")
        public List<String> getStringList() {
            return (List<String>) this.value;
        }

        /**
         * Loads the value from the config and caches it
         */
        private void load(FileConfiguration config) {
            this.value = config.get(this.key);
        }

    }

    @Override
    public void reload() {
        this.plugin.reloadConfig();
        this.plugin.saveDefaultConfig();


        FileConfiguration config = this.plugin.getConfig();
        for (Setting value : Setting.values()) {
            config.addDefault(value.key, value);
            value.load(config);
        }
    }
}

File: src/main/java/xyz/oribuin/skyblock/managers/ConfigManager.kt
package xyz.oribuin.skyblock.managers

import org.bukkit.configuration.file.FileConfiguration
import xyz.oribuin.skyblock.Skyblock

class ConfigManager(plugin: Skyblock) : Manager(plugin) {
    override fun reload() {
        plugin.reloadConfig()
        plugin.saveDefaultConfig()
        val config = plugin.config

        for (value in Setting.values()) {
            config.addDefault(value.key,value.defaultValue)
            value.load(config)
        }

    }

    enum class Setting(val key: String, val defaultValue: Any) {
        WORLD("world-name", "Islands"),
        SETTINGS_SIZE("island-settings.size", 100),
        SQL_ENABLED("my-sql.enabled", false),
        SQL_HOSTNAME("my-sql.hostname", ""),
        SQL_PORT("my-sql.port", 3315),
        SQL_DATABASENAME("my-sql.database-name", ""),
        SQL_USERNAME("my-sql.user-name", ""),
        SQL_PASSWORD("my-sql.password", ""),
        SQL_USE_SSL("my-sql.use-ssl", false);


        private var value: Any? = null

        /**
         * Gets the setting as a boolean
         *
         * @return The setting as a boolean
         */
        val boolean: Boolean
            get() = value as Boolean

        /**
         * @return the setting as an int
         */
        val int: Int
            get() = number.toInt()

        /**
         * @return the setting as a long
         */
        val long: Long
            get() = number.toLong()

        /**
         * @return the setting as a double
         */
        val double: Double
            get() = number

        /**
         * @return the setting as a float
         */
        val float: Float
            get() = number.toFloat()

        /**
         * @return the setting as a String
         */
        val string: String
            get() = value as String

        private val number: Double
            get() {
                if (value is Int) {
                    return (value as Int).toDouble()
                } else if (value is Short) {
                    return (value as Short).toDouble()
                } else if (value is Byte) {
                    return (value as Byte).toDouble()
                } else if (value is Float) {
                    return (value as Float).toDouble()
                }
                return value as Double
            }

        /**
         * @return the setting as a string list
         */
        val stringList: List<*>
            get() = value as List<*>

        /**
         * Loads the value from the config and caches it
         */
        fun load(config: FileConfiguration) {
            value = config[key]
        }

    }

    override fun disable() {
        // Unused
    }
}